File: android/src/main/java/com/reactnativezstd/ZstdModule.java
package com.reactnativezstd;

import android.util.Log;


import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;


import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.ReactMethod;


public class ZstdModule extends ZstdSpec {
  public static final String NAME = "Zstd";
  public static final String CPP_LIB_NAME = "reactnativezstd";

  public ZstdModule(ReactApplicationContext context) {
    super(context);
  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }

  static {
    Log.i(NAME, "Loading C++ library...");
    try {
      System.loadLibrary(CPP_LIB_NAME);
    } catch (Exception e) {
      Log.e(NAME, "Failed to load C++ library", e);
    }
  }


  public static native byte[] nativeCompress(String buffIn,
                                             int compressLevel);

  public static native String nativeDecompress(byte[] buffIn, int sourceSize);


  @ReactMethod(isBlockingSynchronousMethod = true)
  public WritableArray compress(String buffIn, double compressLevel) {
    byte[] compressedData = nativeCompress(buffIn, (int)compressLevel);
    WritableArray compressedDataArray = Arguments.createArray();
    for (byte b : compressedData) {
        compressedDataArray.pushInt(b);
    }
    return compressedDataArray;
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public String decompress(ReadableArray buffIn) {
     byte[] byteArray = new byte[buffIn.size()];
     for (int i = 0; i < buffIn.size(); i++) {
          byteArray[i] = (byte) buffIn.getInt(i);
      }
      return nativeDecompress(byteArray, byteArray.length);
  }


}

File: android/src/main/java/com/reactnativezstd/ZstdModule.kt
package com.reactnativezstd;

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.WritableArray
import com.facebook.react.module.annotations.ReactModule


@ReactModule(name = ZstdModule.NAME)
class ZstdModule(reactContext: ReactApplicationContext) :
  NativeZstdSpec(reactContext) {

  override fun getName(): String {
    return NAME
  }

  companion object {
    const val NAME = "Zstd"

    @JvmStatic
    private external fun nativeCompress(buffIn: String, compressLevel: Int): ByteArray

    @JvmStatic
    private external fun nativeDecompress(byteArray: ByteArray, sourceSize: Int): String

    // Load the native library once when the class is loaded
    init {
      System.loadLibrary("reactnativezstd")
    }
  }

override fun compress(buffIn: String, compressLevel: Double): WritableArray {
    val compressedData = nativeCompress(buffIn, compressLevel.toInt())
    return compressedData.toWritableArray()
}


override fun decompress(buffIn: ReadableArray): String {
    val byteArray = ByteArray(buffIn.size())
    for (i in 0 until buffIn.size()) {
        byteArray[i] = buffIn.getInt(i).toByte()
    }
    return nativeDecompress(byteArray, byteArray.size)
}


private fun ByteArray.toWritableArray(): WritableArray {
      return Arguments.createArray().apply {
          forEach { pushInt(it.toInt()) }
      }
}

}
